_posts/2024-06-14-granite-rag.adoc

```diff
@@ -0,0 +1,286 @@
+
+The RAG pattern operates in two main steps:
+
+1. **Ingestion**: The application ingests a set of documents, processes them, and stores them in a vector database.
+2. **Retrieval**: When a user inputs a query, the application retrieves the most relevant information from the vector database.
+
+The following image gives a high-level overview of the _traditional_ RAG pattern:
+
+image::traditional-rag-pattern.png[The traditional RAG pattern,float="right",align="center"]
+
+There are more advanced version of the RAG pattern, but let's stick to the basics for this application.
+
+=== Ingestion
+
+Let's first look at the ingestion step.
+The ingestion process involves reading a set of documents, splitting them into text segments, computing their vector representations, and storing them in Infinispan.
+
+The secret of an effective RAG implementation lies in how the text segments are computed.
+In our application, we will follow a straightforward approach, but more advanced techniques are available and often required.
+
+The embedding model is responsible for converting text into a vector representation.
+For simplicity, we use an in-process embedding model (`BGE-small`).
+Other options, like the Universal Angle Embedding, are available, but we'll stick with BGE-small for simplicity.
+
+=== Retrieval
+
+The second step is the retrieval process.
+When a user inputs a query, the application searches the vector database to find the most relevant text segments.
+
+To achieve this, the application computes the vector representation of the user query using the **same** embedding model and compares it with the vector representations of the stored text segments in Infinispan.
+It selects the most relevant text segments based on the similarity between the query vector and the text segment vectors.
+
+Then, it augments the user query with the retrieved text segments and sends it to the AI Service.
+Note that until this step, the LLM is not used.
+
+== Implementing the chatbot
+
+Enough theory—let's dive into the implementation.
+You can find the final version on https://github.com/cescoffier/quarkus-granite-rag-demo[GitHub].
+
+=== Used extensions and dependencies
+
+To implement our chatbot, we rely on the following Quarkus extensions:
+
+* `quarkus-langchain4j-openai`: Integrates LLM providers using the OpenAI API, suitable for both Instructlab and Podman AI Studio.
+* `quarkus-websockets-next`: Provides support for WebSocket communication.
+* `quarkus-langchain4j-infinispan`: Integrates Infinispan with Langchain4J, allowing us to store and retrieve vector representations of text segments.
+* `quarkus-web-bundler`: Bundles frontend resources with the Quarkus application.
+
+We also need a specific dependency to use the BGE-small embedding model:
+
+[source,xml]
+----
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-embeddings-bge-small-en-q</artifactId>
+    <version>0.31.0</version>
```

Comment:
You don't the version if you use our BOM

Reply:
So, code.quarkus.io does not add the bom (that's why I have the version) - let me remove it from the blog post but keep it in the code.